### PR TITLE
Lake Formation Data Lake Settings

### PR DESCRIPTION
**Pull request by gmazelier**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #9700

**AWS Lake Formation support:** manages the data lake settings for your account.

This is a proposal for the first resource `LakeFormation::DataLakeSettings` mentioned in #9700, others resources will be provided in upcoming pull requests: `LakeFormation::Permissions`, `LakeFormation::Resource`.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* **New Resource**: `aws_lakeformation_datalake_settings` (#13250)
```

Output from acceptance testing:

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSLakeFormationDataLakeSettings'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSLakeFormationDataLakeSettings -timeout 120m
=== RUN   TestAccAWSLakeFormationDataLakeSettings_basic
=== PAUSE TestAccAWSLakeFormationDataLakeSettings_basic
=== RUN   TestAccAWSLakeFormationDataLakeSettings_withCatalogId
=== PAUSE TestAccAWSLakeFormationDataLakeSettings_withCatalogId
=== CONT  TestAccAWSLakeFormationDataLakeSettings_basic
=== CONT  TestAccAWSLakeFormationDataLakeSettings_withCatalogId
--- PASS: TestAccAWSLakeFormationDataLakeSettings_basic (33.10s)
--- PASS: TestAccAWSLakeFormationDataLakeSettings_withCatalogId (33.50s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	34.906s
```

### CODE REVIEW

File: aws/resource_aws_lakeformation_datalake_settings.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_datalake_settings.go:116]: I suggest renaming to `expandLakeFormationDataLakePrincipal`.

File: aws/resource_aws_lakeformation_datalake_settings.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_datalake_settings.go:129]: I suggest renaming to `flattenLakeFormationDataLakePrincipal`.

File: website/allowed-subcategories.txt
Comment (ewbankkit) [inline website/allowed-subcategories.txt:66]: Nit. `Lake Formation` as that's the [official spelling](https://aws.amazon.com/lake-formation/) of the service.

File: website/aws.erb
Comment (ewbankkit) [inline website/aws.erb:2088]: `Lake Formation` to match `website/allowed-subcategories.txt`.

File: website/docs/r/lakeformation_datalake_settings.html.markdown
Comment (ewbankkit) [inline website/docs/r/lakeformation_datalake_settings.html.markdown:2]: `Lake Formation` to match `website/allowed-subcategories.txt`.

File: aws/resource_aws_lakeformation_datalake_settings.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_datalake_settings.go:30]: I suggest 

```hcl
Type:     schema.TypeSet,
Set:      schema.HashString,
```

here if the order is not important.

File: website/allowed-subcategories.txt
Comment (gmazelier) [reply website/allowed-subcategories.txt]: Updated using official spelling.

File: aws/resource_aws_lakeformation_datalake_settings.go
Comment (gmazelier) [reply aws/resource_aws_lakeformation_datalake_settings.go]: Indeed. Done.

File: aws/resource_aws_lakeformation_datalake_settings.go
Comment (gmazelier) [reply aws/resource_aws_lakeformation_datalake_settings.go]: Renaming done.

File: aws/resource_aws_lakeformation_datalake_settings.go
Comment (gmazelier) [reply aws/resource_aws_lakeformation_datalake_settings.go]: Renaming done.

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSLakeFormationDataLakeSettings_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSLakeFormationDataLakeSettings_ -timeout 120m
=== RUN   TestAccAWSLakeFormationDataLakeSettings_basic
=== PAUSE TestAccAWSLakeFormationDataLakeSettings_basic
=== RUN   TestAccAWSLakeFormationDataLakeSettings_withCatalogId
=== PAUSE TestAccAWSLakeFormationDataLakeSettings_withCatalogId
=== CONT  TestAccAWSLakeFormationDataLakeSettings_basic
=== CONT  TestAccAWSLakeFormationDataLakeSettings_withCatalogId
--- PASS: TestAccAWSLakeFormationDataLakeSettings_basic (25.14s)
--- PASS: TestAccAWSLakeFormationDataLakeSettings_withCatalogId (25.20s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	25.239s
```

File: aws/resource_aws_lakeformation_datalake_settings_test.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_datalake_settings_test.go:28]: Can use `testAccCheckResourceAttrAccountID()` here.

File: aws/resource_aws_lakeformation_datalake_settings_test.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_datalake_settings_test.go:58]: Can use `testAccCheckResourceAttrAccountID()` here.

File: aws/resource_aws_lakeformation_datalake_settings_test.go
Comment (gmazelier) [reply aws/resource_aws_lakeformation_datalake_settings_test.go:28]: You mean checks at l. 27 and l. 28 are redundant? I can remove the last one, IMO there is no need to check `aws_caller_identity` resource alone.

File: aws/resource_aws_lakeformation_datalake_settings_test.go
Comment (gmazelier) [reply aws/resource_aws_lakeformation_datalake_settings_test.go:58]: You mean checks at l. 57 and l. 58 are redundant? Same as above, I can remove the last one, IMO there is no need to check `aws_caller_identity` resource alone.

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Action Required: Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by YakDriver** [comment]
@gmazelier Hello! Are you still interested and available to work on this? We have been unable to get to Lake Formation yet but will be working on it soon.

**Comment by gmazelier** [comment]
@YakDriver Yes, sure. Planned to work on the potential mentioned issues next week.

**Comment by YakDriver** [comment]
@gmazelier I want to be upfront about the process so you know what to expect. A design will be created internally and adjusted and eventually approved. It may seem like nothing is happening at that point but work is happening behind the scenes. Once the design is approved, implementation will need to follow the design. Hopefully we can use some or all of your work. We won't know until we have concensus on the design, however.

**Comment by gmazelier** [comment]
@YakDriver Alright, just let me know if and when I can help.

**Comment by YakDriver** [comment]
@gmazelier My friend, here's the latest. We have the design approved and are moving forward with implementation. To give you a heads up, I will likely be building on your PR. Make sure that you have checked the "allow edits from maintainers" box. You will get credit for the feature and commits. I'll make any changes needed to bring it into alignment with the design.

**Comment by YakDriver** [comment]
@gmazelier Hello, my friend. I have rebased this PR and am aligning it with the design. Thanks for all your hard work and initiative on this. Hopefully I'll complete my end soon. Let me know if you want to commit to your branch since I'm working on it also. Sincères amitiés

**Comment by YakDriver** [comment]
Personally, I like "datalake" as one word but AWS consistently uses "data lake" as two words so I'm renaming to maintain consistency with AWS.

**Comment by gmazelier** [comment]
@YakDriver Sure. Nothing to commit right now, I'm waiting for design realignment. Then I will contribute to provider other resources as you mentioned. But I'll you know. Thank you for your work and feedback! (agree with you about data lake wording but let's be consistent with AWS documentation)

**Comment by YakDriver** [comment]
@ewbankkit @gmazelier Have a look at the latest. If you have any feedback, let me know soon. I'd like to merge today if possible.

### Acceptance tests in `us-west-2`

```
--- PASS: TestAccAWSLakeFormationDataLakeSettings_serial (34.38s)
    --- PASS: TestAccAWSLakeFormationDataLakeSettings_serial/basic (11.86s)
    --- PASS: TestAccAWSLakeFormationDataLakeSettings_serial/disappears (11.46s)
    --- PASS: TestAccAWSLakeFormationDataLakeSettings_serial/withoutCatalogId (11.07s)
```

### Acceptance tests in `us-gov-west-1`

```
--- PASS: TestAccAWSLakeFormationDataLakeSettings_serial (49.85s)
    --- PASS: TestAccAWSLakeFormationDataLakeSettings_serial/basic (17.66s)
    --- PASS: TestAccAWSLakeFormationDataLakeSettings_serial/disappears (16.37s)
    --- PASS: TestAccAWSLakeFormationDataLakeSettings_serial/withoutCatalogId (15.83s)
```

**Review by YakDriver — APPROVED** [review]

**Comment by gmazelier** [comment]
LGTM

**Comment by YakDriver** [comment]
As with any new service, we'll notice things that need adjusting in the first little while. We welcome your help post-release. As for resources and data sources, here is what is planned and not planned. I'll create issues for the items that won't make the first release.

- [ ] resource/lakeformation_resource #13267 
- [ ] data source/lakeformation_resource #13396 
- [ ] data source/lakeformation_resources #13396 (priority 3)
- [ ] resource/lakeformation_data_lake_settings #13250 
- [ ] data source/lakeformation_data_lake_settings #13250 
- [ ] resource/lakeformation_permissions #13396 (priority 1) 
- [ ] data source/lakeformation_permissions #13996 (priority 2)
- [ ] data source/lakeformation_effective_permissions (priority 4, not in 3.22.0)

**Pull request merged**
